### PR TITLE
add --fail-on threshold support

### LINKED ISSUE #129: Create a switch for exiting non-zero if vulnerabilities are found
Users are going to want to use `grype` in their CI/CD pipelines. Typically, this will occur right after they've built a container image.

_Pipelines rely on the exit status of processes to know whether the pipeline has "failed" or should continue with subsequent steps. If a process (like a linter, a compiler, a test runner, etc.) exits 1 (or some non-zero number), that communicates that this particular CI job has failed._

With `grype`, users should be able to ask that the process exit non-zero when vulnerabilities are found, so that their pipelines fail — intentionally.

Example:

```
$ grype --enforce my-image
... a list of vulnerabilities ...
$ echo $?
1
```

The exact name of the switch itself is up for discussion, but should be both concise and easily understood by others.

It's important that this behavior be *opt-in*. By default, `grype` should exit non-zero only when a runtime error has occurred. There are plenty of scenarios in which a user would expect `grype` to exit `0` even when vulnerabilities are present (for example, when the output of grype is being piped to a downstream process).

### PR DESCRIPTION
**Pull request by wagoodman**
Adds the following functionalities:
- the ability to comb through the discovered vulnerabilities and set a non-0 exit code when a vulnerability at or above the given user threshold is found (provided by `--fail-on`).
- a `ExpectedErr` struct for creating errors that are known to be produced from grype. There is room to expand this in the future.

Note: the drop in coverage is somewhat fictitious since a test was added in a package that has no existing tests. 

Closes #129

### CODE REVIEW

File: grype/grypeerr/errors.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package grypeerr
```
Comment (dakaneye) [inline grype/grypeerr/errors.go:1]: Nit: Would it be worth hyphenating this to make it more readable? I know the k8s client has a hyphen in the package name, so should be ok?

File: cmd/root.go [hunk 2]
```diff
@@ -179,12 +204,34 @@ func startWorker(userInput string) <-chan error {
...
-	errs := startWorker(userInput)
+	errs := startWorker(userInput, appConfig.FailOnSeverity)
 	ux := ui.Select(appConfig.CliOptions.Verbosity > 0, appConfig.Quiet)
 	return ux(errs, eventSubscription)
 }
 
-func ListLocalDockerImages(prefix string) ([]string, error) {
+// hitSeverityThreshold indicates if there are any severities >= to the max allowable severity (which is optional)
+func hitSeverityThreshold(thresholdSeverity *vulnerability.Severity, results result.Result, metadataProvider vulnerability.MetadataProvider) bool {
+	if thresholdSeverity != nil {
```
Comment (dakaneye) [inline cmd/root.go:214]: could reduce the nesting in this method a bit:
```
if thresholdSeverity == nil {
    return true
}
```

**Review by dakaneye — APPROVED** [review]

File: grype/grypeerr/errors.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+
+var (
+	// ErrAboveSeverityThreshold indicates when a vulnerability severity is discovered that is above the given --fail-on severity value
+	ErrAboveSeverityThreshold = NewExpectedErr("discovered vulnerabilities at or above the severity threshold")
```
Comment (luhring) [inline grype/grypeerr/errors.go:5]: I would probably move specific errors out of the general error package and closer to where the error is relevant. For example, if the severity checking logic is in the `cmd` package right now, that's where I'd put this error.

File: grype/grypeerr/errors.go (hunk 1 above)
Comment (luhring) [reply grype/grypeerr/errors.go:1]: It's okay for hyphens to appear in module paths, but not package names. If we inserted a hyphen so that we have `package grype-err`, the code wouldn't compile.

[Effective Go](https://golang.org/doc/effective_go.html#package-names) (another great read for aspiring Go devs, btw!) puts forth the convention of packages that use "lower case, single-word names; there should be no need for underscores or mixedCaps". I don't think we need to apply this rule to 100% of cases, but it's worth keeping in mind.

In short, I'm okay with leaving the code as written. If we want to consider alternative approaches, one might be to name this package just "err" or something. So full path would be `github.com/anchore/grype/grype/err`. This would likely lead to aliasing the package when imported throughout our code, for example:

```go
import (
    grypeerr "github.com/anchore/grype/grype/err"
)
```

So I'm not sure it's a net win to change it.

File: cmd/root.go (hunk 2 above)
Comment (luhring) [reply cmd/root.go:214]: Agreed. Also, I think this could potentially be further simplified by having the severity variable not be a pointer type.

File: internal/ui/etui/ephemeral_tui.go [hunk 4]
```diff
@@ -73,16 +75,25 @@ func OutputToEphemeralTUI(workerErrs <-chan error, subscription *partybus.Subscr
 	ctx := context.Background()
 	grypeUIHandler := grypeUI.NewHandler()
 
-eventLoop:
+	var errResult error
 	for {
 		select {
-		case err := <-workerErrs:
+		case err, ok := <-workerErrs:
```
Comment (luhring) [inline internal/ui/etui/ephemeral_tui.go:81]: GoLand underlined `workerErrs` with this warning: "Receive may block because of 'nil' channel". May be worth checking for `nil` ahead of time.

File: internal/ui/etui/ephemeral_tui.go [hunk 5]
```diff
@@ -73,16 +75,25 @@ func OutputToEphemeralTUI(workerErrs <-chan error, subscription *partybus.Subscr
...
+					errResult = err
+					continue
+				}
 				return err
 			}
+			if !ok {
+				// worker completed
+				workerErrs = nil
+			}
 		case e, ok := <-events:
```
Comment (luhring) [inline internal/ui/etui/ephemeral_tui.go:93]: GoLand underlined `events` with this warning: "Receive may block because of 'nil' channel". May be worth checking for `nil` ahead of time.

File: internal/ui/etui/ephemeral_tui.go [hunk 6]
```diff
@@ -73,16 +75,25 @@ func OutputToEphemeralTUI(workerErrs <-chan error, subscription *partybus.Subscr
...
 
-eventLoop:
+	var errResult error
 	for {
 		select {
-		case err := <-workerErrs:
+		case err, ok := <-workerErrs:
 			if err != nil {
+				if errors.Is(err, grypeerr.ErrAboveSeverityThreshold) {
+					errResult = err
```
Comment (luhring) [inline internal/ui/etui/ephemeral_tui.go:84]: Definitely up for discussion --> if I'm reading this correctly, it looks like the logic around `errResult` is such that errors that aren't the latest error are lost forever. Is that true? If so, would it be worth collecting surfaced errors for reporting, if we're not displaying them immediately?

File: internal/ui/etui/ephemeral_tui.go [hunk 7]
```diff
@@ -112,15 +123,15 @@ eventLoop:
...
 				// this is the last expected event
-				break eventLoop
+				events = nil
 			}
 		case <-ctx.Done():
-			if ctx.Err() != nil {
-				log.Errorf("cancelled (%+v)", err)
-			}
-			break eventLoop
+			return grypeerr.NewExpectedErr("cancelled: %w", ctx.Err())
```
Comment (luhring) [inline internal/ui/etui/ephemeral_tui.go:129]: Nit: sp: "canceled"

File: grype/grypeerr/errors.go (hunk 1 above)
Comment (wagoodman) [reply grype/grypeerr/errors.go:1]: I agree with both of you --I don't like `grypeerr` but also agree with Dan's points and don't see a better alternative. The `err` name is problematic since, when imported, would be overridden by the most popular local variable name for errors within functions. I'll leave it as is for now.

File: cmd/root.go (hunk 2 above)
Comment (wagoodman) [reply cmd/root.go:214]: 👍 guard clases are best!

File: grype/grypeerr/errors.go (hunk 3 above)
Comment (wagoodman) [reply grype/grypeerr/errors.go:5]: Agreed, I'll move this over to `cmd`

File: internal/ui/etui/ephemeral_tui.go (hunk 5 above)
Comment (wagoodman) [reply internal/ui/etui/ephemeral_tui.go:93]: per our conversation, this section probably needs more love than this PR is willing to give

File: internal/ui/etui/ephemeral_tui.go (hunk 6 above)
Comment (wagoodman) [reply internal/ui/etui/ephemeral_tui.go:84]: this is here to prevent return of an error and short circuit the UI logic --per our conversation, this section needs a second pass on design. It is OK that we are capturing the most recent `errResult` since this is only there to indicate `ErrAboveSeverityThreshold` --will not scale though.

File: internal/ui/etui/ephemeral_tui.go (hunk 4 above)
Comment (wagoodman) [reply internal/ui/etui/ephemeral_tui.go:81]: [ditto](https://github.com/anchore/grype/pull/156#discussion_r492236148)

File: grype/grypeerr/errors.go (hunk 3 above)
Comment (wagoodman) [reply grype/grypeerr/errors.go:5]: per our conversation, let's iterate on this --we should try not to have a global sentinel errors package, but probably isn't a show stopper for now (especially since we don't have a better pattern off hand yet)
